internal/engine/crd.go

```diff
@@ -0,0 +1,152 @@
+	Original cue.Value
+	Props    struct {
+		Spec struct {
+			Group string `json:"group"`
+			Names struct {
+				Kind     string `json:"kind"`
+				ListKind string `json:"listKind"`
+				Plural   string `json:"plural"`
+				Singular string `json:"singular"`
+			} `json:"names"`
+		} `json:"spec"`
+	}
+
+	// All the schemas in the original CRD, converted to CUE representation.
+	Schemas []VersionedSchema
+}
+
+// VersionedSchema is an intermediate form of a single versioned schema from a CRD
+// (an element in `spec.versions`), converted to CUE.
+type VersionedSchema struct {
+	// The contents of the `spec.versions[].name`
+	Version string
+	// The contents of `spec.versions[].schema.openAPIV3Schema`, after conversion of the OpenAPI
+	// schema to native CUE constraints.
+	Schema cue.Value
+}
+
+func convertCRD(crd cue.Value) (*IntermediateCRD, error) {
+	cc := &IntermediateCRD{
+		Schemas: make([]VersionedSchema, 0),
+	}
+
+	err := crd.Decode(&cc.Props)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding crd props into Go struct: %w", err)
+	}
+
+	vlist := crd.LookupPath(cue.ParsePath("spec.versions"))
+	if !vlist.Exists() {
+		return nil, fmt.Errorf("crd versions list is absent")
+	}
+	iter, err := vlist.List()
+	if err != nil {
+		return nil, fmt.Errorf("crd versions field is not a list")
+	}
+
+	ctx := crd.Context()
+	shell := ctx.CompileString(`
+		openapi: "3.0.0",
+		info: {
+			title: "dummy",
+			version: "1.0.0",
+		}
+		components: schemas: thedef: _
+	`)
+	schpath := cue.ParsePath("components.schemas.thedef")
+	defpath := cue.MakePath(cue.Def("thedef"))
+
+	// The CUE stdlib openapi encoder expects a whole openapi document, and then
+	// operates on schema elements defined within #/components/schema. Each
+	// versions[].schema.openAPIV3Schema within a CRD is ~equivalent to a single
+	// element under #/components/schema, as k8s does not allow CRD schemas to
+	// contain any kind of external references.
+	//
+	// So, for each schema.openAPIV3Schema, we wrap it in an openapi document
+	// structure, convert it to CUE, then merge it back into place at its
+	// original position.
+	var i int
+	for iter.Next() {
+		val := iter.Value()
+		ver, err := val.LookupPath(cue.ParsePath("name")).String()
+		if err != nil {
+			return nil, fmt.Errorf("unreachable? error getting version field for versions element at index %d: %w", i, err)
+		}
+		i++
+
+		doc := shell.FillPath(schpath, val.LookupPath(cue.ParsePath("schema.openAPIV3Schema")))
+		of, err := openapi.Extract(doc, &openapi.Config{})
+		if err != nil {
+			return nil, fmt.Errorf("could not convert schema for version %s to CUE: %w", ver, err)
+		}
+		sch := ctx.BuildFile(of)
+		cc.Schemas = append(cc.Schemas, VersionedSchema{
+			Version: ver,
+			Schema:  sch.LookupPath(defpath),
+		})
+
+		// Additional massaging of converted schemas should be done here. Lots
+		// could be done. One example: apiVersion field appears to be coming
+		// through conversion as optional.
```

Comment:
i pretty heavily expect this to be something that we'll want to do.

For example, to handle erroneous field optionality (and probably some other issues i haven't yet noticed), the simplest thing would be to get some native CUE definitions of what all CRDs must look like committed to timoni (similar to [this](https://github.com/grafana/kindsys/blob/74f2bc8126104cd5c989a8d0e2a8539f6a8a8b2e/kindcat_custom.cue#L41-L97)).

Then, in this func, we'd unify those static definitions with the one being processed. As long as `apiVersion` is required in the static definition, it won't matter whether it comes through from openapi extraction as required or not.

Reply:
Here I would drop all fields except for `spec` and `status`, and inject the Kubernetes Object props:

```
apiVersion: *"source.toolkit.fluxcd.io/v1beta2" | string
kind:       *"Bucket" | string

metadata: {
    name:       string
    namespace?: string
    labels?: {[string]: string}
    annotations?: {[string]: string}
}
```

Timoni users shouldn't have to set the `apiVersion` nor `kind` in their objects, as these values are in the original CRD.

Reply:
yep, all seems good, done